### PR TITLE
Allow Transactions with Null Wallets

### PR DESCRIPTION
**Pull request by leonardocustodio**
## PR Type:
Refactoring

___
## PR Description:
This PR introduces changes that allow transactions to be created with null wallets. The changes are reflected in the tests and the database schema. The main changes include:
- Updated various test files to replace wallet details with null in the transaction creation process.
- Updated the TransactionService.php file.
- Updated the HasSigningAccountField.php file.
- A new migration file is added to make the account nullable in transactions.
- The TransactionCreated event and the CoreServiceProvider have been updated.

___
## PR Main Files Walkthrough:
<details> <summary>files:</summary>

- `tests/Feature/GraphQL/Mutations/BatchMintTest.php`: Replaced wallet details with null in the transaction creation process in various test cases.
- `tests/Feature/GraphQL/Mutations/CreateTokenTest.php`: Replaced wallet details with null in the transaction creation process in various test cases.
- `tests/Feature/GraphQL/Mutations/BatchTransferTest.php`: Replaced wallet details with null in the transaction creation process in various test cases.
- `tests/Feature/GraphQL/Mutations/MintTokenTest.php`: Replaced wallet details with null in the transaction creation process in various test cases.
- `src/Services/Database/TransactionService.php`: The file has been updated but the specific changes are not visible in the provided diff.
- `src/GraphQL/Types/Input/Substrate/Traits/HasSigningAccountField.php`: The file has been updated but the specific changes are not visible in the provided diff.
- `database/migrations/2023_11_06_133605_make_account_nullable_in_transactions.php`: A new migration file to make the account nullable in transactions.
- `src/Events/Global/TransactionCreated.php`: The file has been updated but the specific changes are not visible in the provided diff.
- `src/CoreServiceProvider.php`: The file has been updated but the specific changes are not visible in the provided diff.
</details>

### CODE REVIEW

File: src/GraphQL/Types/Input/Substrate/Traits/HasSigningAccountField.php [hunk 1]
```diff
@@ -30,10 +30,14 @@ public function getSigningAccountField(
...
      * Get the signing account.
      */
-    public function getSigningAccount(array $args)
+    public function getSigningAccount(array $args): ?Model
     {
-        return WalletService::firstOrStore([
-            'account' => empty($signing = Arr::get($args, 'signingAccount')) ? Account::daemonPublicKey() : $signing,
-        ]);
+        if (!empty($signing = Arr::get($args, 'signingAccount'))) {
+            return WalletService::firstOrStore([
```
Comment (enjinabner) [inline src/GraphQL/Types/Input/Substrate/Traits/HasSigningAccountField.php:36]: I think there's a problem here, how will the signingAccount be populated with multi-tenant? I don't think the wallet daemon knows what's the wallet address of the user

I think we need to fix this part in `Account.php` instead

```
public static function daemonPublicKey(): ?string
{
        return static::$publicKey ? SS58Address::getPublicKey(static::$publicKey) : null;
}
```

Take note that for multi-tenant we are setting this `static::$publicKey` attribute here https://github.com/enjin/platform-multi-tenant/blob/feature/PLA-845/multi-tenant-system/src/Services/Auth/Drivers/SanctumAuth.php#L37


```suggestion
       $account =  empty($signing = Arr::get($args, 'signingAccount')) ? Account::daemonPublicKey() : $signing
       if ($account) {
              return WalletService::firstOrStore(['account' => $account]);
       }
       return null;
```

File: src/GraphQL/Schemas/Primary/Mutations/MarkAndListPendingTransactionsMutation.php [hunk 2]
```diff
@@ -77,7 +77,9 @@ function (Builder $query) use ($args) {
                     return $query->whereIn('wallet_public_key', $publicKeys);
                 },
                 function (Builder $query) {
-                    return $query->whereIn('wallet_public_key', Account::managedPublicKeys());
+                    return $query
```
Comment (enjinabner) [inline src/GraphQL/Schemas/Primary/Mutations/MarkAndListPendingTransactionsMutation.php:80]: I think query will look like this and can cause unexpected result
```select * from transactions where state = 'PENDING' and wallet_public_key in('public keys'...) OR wallet_public_key = null```

the correct query should be 
```select * from transactions where state = 'PENDING' and (wallet_public_key in('public keys'...) OR wallet_public_key is null)```


```suggestion
            return $query->where(function($query) {
                   ->whereIn('wallet_public_key', Account::managedPublicKeys())
                   ->orWhereNull('wallet_public_key');
            });
```
FYI, i haven't tested this but code should be something like this

**Review by enjinabner — CHANGES_REQUESTED** [review]

File: src/GraphQL/Types/Input/Substrate/Traits/HasSigningAccountField.php (hunk 1 above)
Comment (leonardocustodio) [reply src/GraphQL/Types/Input/Substrate/Traits/HasSigningAccountField.php:36]: I'm not sure I got your point here. The signing account is an address, the address is unique. We don't have the same address for different users. Also each user has its own daemon, we don't need to know the user address.

Comment (leonardocustodio) [reply src/GraphQL/Types/Input/Substrate/Traits/HasSigningAccountField.php:36]: tagging @v16Studios as he is more used with the multi-tenant package.

File: src/GraphQL/Schemas/Primary/Mutations/MarkAndListPendingTransactionsMutation.php (hunk 2 above)
Comment (enjinabner) [reply src/GraphQL/Schemas/Primary/Mutations/MarkAndListPendingTransactionsMutation.php:80]: Anyway i think query is still wrong, we shouldn't do `wallet_public_key = null` for null comparison instead it should be `wallet_public_key is null`

https://dev.mysql.com/doc/refman/8.0/en/working-with-null.html

Comment (leonardocustodio) [reply src/GraphQL/Schemas/Primary/Mutations/MarkAndListPendingTransactionsMutation.php:80]: Well we are using laravel:
```
        // If the value is "null", we will just assume the developer wants to add a
        // where null clause to the query. So, we will allow a short-cut here to
        // that method for convenience so the developer doesn't have to check.
        if (is_null($value)) {
            return $this->whereNull($column, $boolean, $operator !== '=');
        }
```

Comment (enjinabner) [reply src/GraphQL/Schemas/Primary/Mutations/MarkAndListPendingTransactionsMutation.php:80]: Good to know

**Review by enjinabner — APPROVED** [review]

**Pull request merged**